refactor: return MempoolAcceptResult from ATMP

glozow opened the pull request:
This is the first 4 commits of #20833, and does refactoring only. It should be relatively simple to review, and offers a few nice things:
- It makes accessing values that don't make sense (e.g. fee) when the tx is invalid an error.
- Returning `MempoolAcceptResult` from ATMP makes the interface cleaner. The caller can get a const instead of passing in a mutable "out" param.
- We don't have to be iterating through a bunch of lists for package validation, we can just return a `std::vector<MempoolAcceptResult>`.
- We don't have to refactor all ATMP call sites again if/when we want to return more stuff from it.

glozow committed:
[refactor] return MempoolAcceptResult

This creates a cleaner interface with ATMP, allows us to make results const,
and makes accessing values that don't make sense (e.g. fee when tx is
invalid) an error.

glozow committed:
[refactor] const ATMPArgs and non-const Workspace

ATMPArgs should contain const arguments for validation.
The Workspace should contain state that may change
throughout validation.

glozow committed:
[refactor] clean up logic in testmempoolaccept

Cleans up reundant code and reduces the diff of the next commit.

glozow committed:
[refactor] improve style for touched code

ariard commented on src/rpc/rawtransaction.cpp:
```diff
@@ -946,32 +946,27 @@ static RPCHelpMan testmempoolaccept()
+        // Check that fee does not exceed maximum fee
+        if (max_raw_tx_fee && fee > max_raw_tx_fee) {
+            result_0.pushKV("allowed", false);
+            result_0.pushKV("reject-reason", "max-fee-exceeded");
```
abc6ff1

I don't know about returning "allowed"=false for `max_raw_tx_fee` violation. "allowed" is documented as "If the mempool allows this tx to be inserted". Your transaction might be mempool valid but doesn't pass the client belt-and-suspender, a different check. 

Also it would be nice to return "base" and "fees" to let the transaction construction builder adapt the feerate to something passing `max_raw_tx_fee`.

Note, that's independent from refactoring, more a undersight of #19339.

glozow replied:
I agree 😢 I was a bit naive when I wrote that. But it's not part of this PR, so maybe we can discuss elsewhere

ariard replied:
Tracked #21074

ariard commented on src/validation.cpp:
```diff
@@ -380,10 +380,8 @@ static void UpdateMempoolForReorg(CTxMemPool& mempool, DisconnectedBlockTransact
         if (!fAddToMempool || (*it)->IsCoinBase() ||
-            !AcceptToMemoryPool(mempool, stateDummy, *it,
-                                nullptr /* plTxnReplaced */, true /* bypass_limits */)) {
+            !AcceptToMemoryPool(mempool, *it, true /* bypass_limits */).m_accepted.value()) {
```
abc6ff1

I don't think that's a good idea to encumber code path like `UpdateMempoolForReorg` with `std::optional`. If this `nullopt` will throw an exception. And we do have the failure/unfinished constructor allowing `m_accepted` to be initialized to nullopt, even if AFAICT such constructor is never called with `finished=bad`for now ? 

I know there is a discussion about `std::optional` usage [here](https://github.com/bitcoin/bitcoin/pull/20833#discussion_r557185885) but could we restrain its usage to only `m_base_fee`/`m_replaced_transaction` ? They would be `nullopt` if `m_accepted=false`.

glozow replied:
Check now? I think it's better with enum

ariard replied:
Do we really need an enum and can't we rely only on a boolean ? Maybe you can point me to a branch how you're using UNFINISHED because we don't use it with this PR ?

glozow replied:
Haven't published the branch yet but the idea is to return a `std::vector<MempoolAcceptResult>` from `ProcessNewPackage`, quit early when a tx fails, and set the not-fully-validated txns in the package to `UNFINISHED`. See [comment](https://github.com/bitcoin/bitcoin/pull/20833#discussion_r567952639)?

ariard replied:
I think `UNFINISHED`/not_fully_validated to mark package partial failure doesn't bring further value compared to unvalid. Do you plan to consume this `UNFINISHED` in a special way ? Otherwise we can just extend `TxValidationResult` with a `PACKAGE_PARENT_FAILED` instead of yet-another-state variable.

glozow replied:
I think in the future with Package Relay, we may want to punish nodes differently for invalidity in packages vs invalidity in transactions. Similarly, it may affect how we cache failed transactions. Obviously this isn't set in stone, but I think it would be better no to put package-specific validation info in `TxValidationState` - what do you think?

ariard replied:
I don't see a straightforward reason to punish faulty pacakge-relay peers from regular tx-relay ones, at least at the mempool level. If we don't have a motivation for `UNFINISHED`, let's remove it for now and defer its introduction when we actually hit the case ? Would be easier to argue at that point.

ariard commented on src/validation.cpp:
```diff
@@ -1007,8 +1002,7 @@ bool MemPoolAccept::Finalize(ATMPArgs& args, Workspace& ws)
                 (int)entry->GetTxSize() - (int)nConflictingSize);
-        if (args.m_replaced_transactions)
-            args.m_replaced_transactions->push_back(it->GetSharedTx());
+        args.m_replaced_transactions.emplace_back(it->GetSharedTx());
```
abc6ff1

Note, you might have a slight performance regression due to `emplace-back` memory allocation if container capacity is too small. Before this change, this potential reallocation wasn't the default.

I think that's okay `replaced_transactions` was already existent in the most busiest code path `ProcessMessage`.

glozow replied:
switched it back to `push_back` to adhere to my promise of "pure refactor"

ariard replied:
I think switching to `push_back` still do an automatic reallocation of the allocated storage if container capacity is too short. See https://www.cplusplus.com/reference/vector/vector/emplace_back/ and https://www.cplusplus.com/reference/vector/vector/push_back/.

My point was underscoring that the allocation was previously conditional of `m_replaced_transaction` pointer being opt-in by ATMP caller. After this change it's not anymore. Again I don't think that matters that most.

ariard commented on src/validation.h:
```diff
@@ -181,12 +181,41 @@ void UnlinkPrunedFiles(const std::set<int>& setFilesToPrune);
+};
+
+/**
+ * (Try to) add a transaction to the memory pool.
```
abc6ff1

"Evaluate a transaction mempool acceptance. If `test_accept`= false, add the transaction to the mempool".

glozow replied:
Added!

ariard replied:
To be precise, I think `bypass_limits` also overrides some feerate check (L729 in `src/validation.cpp`).

glozow replied:
Addressed

ariard commented on src/validation.h:
```diff
@@ -181,12 +181,41 @@ void UnlinkPrunedFiles(const std::set<int>& setFilesToPrune);
+* of replaced transactions and m_base_fees contains the tx fees.
+* When m_accepted = nullopt, the transaction was not fully validated
+* (may be valid or invalid), e.g. when it is part of a package that
+* failed validation for an earlier transaction.
```
abc6ff1

See other comment about `std::optional` usage, but at least indicate what does `m_accept = false` mean.

ariard commented on src/validation.h:
```diff
@@ -181,12 +181,41 @@ void UnlinkPrunedFiles(const std::set<int>& setFilesToPrune);
+
+    // Valid when m_accepted = true
+    std::optional<std::list<CTransactionRef>> m_replaced_transactions;
+    std::optional<CAmount> m_base_fees;
```
abc6ff1

As this interface aims to be consumed by higher users, maybe comment those variables, like "evicted transactions per BIP 125 rules", "Raw base fees expressed in satoshis".

glozow replied:
Added comments

ariard replied:
Can you added the unit in comment (satoshi). I swear once you start to deal with fee computation you spend your time thinking if you're dealing with apples or oranges :)

ariard commented on src/validation.h:
```diff
@@ -181,12 +181,41 @@ void UnlinkPrunedFiles(const std::set<int>& setFilesToPrune);
+* Validation result for a single transaction mempool acceptance.
+* When m_accepted = true, m_replaced_transactions contains a list
+* of replaced transactions and m_base_fees contains the tx fees.
+* When m_accepted = nullopt, the transaction was not fully validated
```
abc6ff1

I think we should commit to a clearer terminology. Validity is always wholesome, a transaction or package is either valid or not. But validity is function of a mempool acceptance evaluation and this is configurable (`bypass_limit`), stateful (e.g mempool min feerate), depends if the transaction is part of a package, etc.

If we follow this line, maybe we should rename `m_accepted` to `m_valid` and have only binary state (true, false). The interface could be extended in the future to indicate it's part of a package, and we may have a TxValidationState for package children to inherit the invalidity (`PACKAGE_INVALID_PARENT`).

ariard reviewed (commented):
Thanks for splitting this from #20833!

glozow commented on src/validation.cpp:
```diff
@@ -1031,46 +1025,51 @@ bool MemPoolAccept::Finalize(ATMPArgs& args, Workspace& ws)
     // Tx was accepted, but not added
-    if (args.m_test_accept) return true;
+    if (args.m_test_accept) {
+        return MempoolAcceptResult(args.m_state, std::move(args.m_replaced_transactions), args.m_fee_out);
```
What do you think of changing `m_accepted` to an enum to encompass states beyond valid/invalid? Something like 

```c
enum class ResultType : uint8_t {
        UNSET, //!> Not fully validated, quit early for whatever reason.
        INVALID, //!> Invalid.
        VALID, //!> Valid.
        VALID_BYPASSED, //!> bypass_limits = true
}
```
This would leave room for a `ResultType::VALID_BYPASSED_TIMELOCKS`

ariard replied:
See my other comment but I would keep a binary state for validity. I don't think `TxValidationResult`/`BlockValidationResult` are great examples, it makes it harder to reason on once you multiply states. 

Let's keep this suggestion in mind for now, it's not a must for this PR. We'll see if we need to introduce something like this if we have situation where we have one aware caller and multiple blind consumers.

darosior reviewed (commented):
Concept ACK -will review soon

glozow commented:
Teeny rebase for the compiler warnings and changed from `std::optional<bool> m_accepted` to an enum `ResultType m_result_type` so there's no risk of throwing for bad optional access.

jnewbery commented on src/rpc/rawtransaction.cpp:
```diff
@@ -946,44 +946,35 @@ static RPCHelpMan testmempoolaccept()
+            fees.pushKV("base", ValueFromAmount(fee));
+            result_0.pushKV("fees", fees);
+        }
+        result.push_back(std::move(result_0));
```
`result.push_back(std::move(result_0));` is the last line in both branches of the if/else. Would it be better to leave it outside the if/else?

jnewbery replied:
I guess this is because it's better to structure like this for package acceptance.

glozow replied:
It makes more sense with the package all valid / 1 invalid tx branches, but not as much here. I suppose leaving it like this reduces the diff for later

jnewbery commented on src/validation.cpp:
```diff
@@ -688,10 +686,7 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
-    if (args.m_fee_out) {
-        *args.m_fee_out = nFees;
-    }
+    ws.m_base_fees = nFees;
```
Do we even need the local `nFees` variable, now that `ws.m_base_fees` is always set? Can you just pass a reference to `ws.m_base_fees` to `CheckInputs()`?

glozow replied:
True, done

jnewbery commented on src/validation.h:
```diff
@@ -181,12 +181,47 @@ void UnlinkPrunedFiles(const std::set<int>& setFilesToPrune);
+
+    // The following fields are only present when m_result_type = ResultType::VALID
+    /** Mempool transactions replaced by the tx per BIP 125 rules. */
+    std::optional<std::list<CTransactionRef>> m_replaced_transactions;
```
I don't think the `std::optional` wrapper is adding anything here. Essentially all it does is add a boolean where false means the value is undefined and true means that it's defined. You're already saying that these fields are only defined if `m_result_type = ResultType::VALID`, so unless you think there's some reason to add this for safety or clarity, I wouldn't bother.

glozow replied:
I think there should be a distinction between "tx didn't replace any txns" and "tx was invalid and thus couldn't replace any txns," and it should be an error to try to look at `m_replaced_transactions` if the tx was invalid.

jnewbery replied:
ok

jnewbery commented on src/validation.h:
```diff
@@ -181,12 +181,47 @@ void UnlinkPrunedFiles(const std::set<int>& setFilesToPrune);
+    * including the possibility of unfinished validation.
+    */
+    enum class ResultType : uint8_t {
+        UNFINISHED, //!> Not fully validated.
```
The tristate here could be confusing for reviewers without the context of package acceptance. Perhaps this comment could be expanded a little. Something like:

```suggestion
        UNFINISHED, //!> Not fully validated. Only used for package mempool acceptance where an individual tx may not be fully validated if the package fails.
```

Also consider adding an out-assertion to `AcceptToMemoryPool()`:

```
MempoolAcceptResult AcceptToMemoryPool(CTxMemPool& pool, const CTransactionRef &tx, bool bypass_limits, bool test_accept)
{
    auto res = AcceptToMemoryPoolWithTime(Params(), pool, tx, GetTime(), bypass_limits, test_accept);
    // For single tx acceptance, validation must not be partial.
    assert(res.m_result_type == MempoolAcceptResult::ResultType::VALID ||
           res.m_result_type == MempoolAcceptResult::ResultType::INVALID);  
    return res;

```

glozow replied:
Resolved by removing `UNFINISHED`, I will plan to add this assertion whenever we have more than 2 result types 👍

jnewbery commented on src/validation.h:
```diff
@@ -181,12 +181,47 @@ void UnlinkPrunedFiles(const std::set<int>& setFilesToPrune);
+
+    /** Constructor for failure or unfinished case */
+    explicit MempoolAcceptResult(TxValidationState state, bool finished=true) :
+        m_state(state), m_replaced_transactions(nullopt), m_base_fees(nullopt) {
+            m_result_type = finished ? ResultType::INVALID : ResultType::UNFINISHED;
```
You can set m_result_type in the initializer list:

```suggestion
        m_result_type(finished ? ResultType::INVALID : ResultType::UNFINISHED),
        m_state(state), m_replaced_transactions(nullopt), m_base_fees(nullopt) {}
```

glozow replied:
(Will do this 👀 )

jnewbery reviewed (commented):
Code review ACK bc992b7a394629137929647998149f18fea5ab29

A few non-essential style suggestions inline. You're going to be touching this code again in the next PRs, so feel free to defer these suggestions.

glozow commented:
Addressed @jnewbery comments and @ariard https://github.com/bitcoin/bitcoin/pull/21062#discussion_r569330388 and removed the `ResultType::UNFINISHED` for now since it's unused. Leaving it as an enum because I think we agree that there's value in having more than 2 states, and the struct is the same size

jnewbery commented:
Code review ACK a9ff9c1ca0530e448341e9d24ecd5f8bc6f2ee42

Cirrus failure looks spurious, but I don't know how to restart it.

MarcoFalke commented on src/validation.h:
```diff
@@ -181,12 +181,45 @@ void UnlinkPrunedFiles(const std::set<int>& setFilesToPrune);
+    /** Constructor for success case */
+    explicit MempoolAcceptResult(TxValidationState state,
+                                 std::list<CTransactionRef>&& replaced_txns, CAmount fees) :
+        m_result_type(ResultType::VALID), m_state(state),
```
What is the point of setting the state when the default constructed one is already valid? It seems like a footgun to allow the caller to pass an invalid state here. At least, it should `ASSUME(state.IsValid());`.

Also, clang-format will like you more when you put the `:` in the next line:

```
    : m_result_type{...

glozow replied:
You're right, there's no point in passing the state when it's successful

MarcoFalke commented on src/validation.h:
```diff
@@ -181,12 +181,45 @@ void UnlinkPrunedFiles(const std::set<int>& setFilesToPrune);
+    /** Used to indicate the results of mempool validation,
+    * including the possibility of unfinished validation.
+    */
+    enum class ResultType : uint8_t {
```
nit: If ResultType is never serialized or used as an enum flag, you don't need to specify the underlying type. And I think the compiler will fill this to at least 32 bits inside the struct anyway, so it won't give you any space savings either.

MarcoFalke commented on src/validation.cpp:
```diff
@@ -1031,46 +1025,51 @@ bool MemPoolAccept::Finalize(ATMPArgs& args, Workspace& ws)
-    MemPoolAccept::ATMPArgs args { chainparams, state, nAcceptTime, plTxnReplaced, bypass_limits, coins_to_uncache, test_accept, fee_out };
-    bool res = MemPoolAccept(pool).AcceptSingleTransaction(tx, args);
-    if (!res) {
+    MemPoolAccept::ATMPArgs args { chainparams, state, nAcceptTime, {}, bypass_limits, coins_to_uncache, test_accept, {} };
```
The two `{}`, as well as `state` aren't args, but workspace variables. Is there any reason to not put them in the workspace?

MarcoFalke replied:
Ok, I see this is fixed in the next commit

MarcoFalke commented on src/validation.cpp:
```diff
@@ -489,14 +486,17 @@ class MemPoolAccept
+        std::list<CTransactionRef> m_replaced_transactions;
 
         bool m_replacement_transaction;
+        CAmount m_fee_out;
```
This should say m_base_fees


```
validation.cpp:684:103: error: no member named 'm_base_fees' in '(anonymous namespace)::MemPoolAccept::Workspace'
    if (!Consensus::CheckTxInputs(tx, state, m_view, g_chainman.m_blockman.GetSpendHeight(m_view), ws.m_base_fees)) {
                                                                                                   ~~ ^
validation.cpp:702:24: error: no member named 'm_base_fees' in '(anonymous namespace)::MemPoolAccept::Workspace'
    nModifiedFees = ws.m_base_fees;
                    ~~ ^
validation.cpp:716:45: error: no member named 'm_base_fees' in '(anonymous namespace)::MemPoolAccept::Workspace'
    entry.reset(new CTxMemPoolEntry(ptx, ws.m_base_fees, nAcceptTime, ::ChainActive().Height(),
                                         ~~ ^
3 errors generated.

glozow replied:
Agh 🤦 bad rebase.

maflcko reviewed (dismissed):
The third commit doesn't compile

e2713614089f77a01c98bb6783f2b1a22aa6ab13 🎪

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

e2713614089f77a01c98bb6783f2b1a22aa6ab13 🎪
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUidXwwAiGoRXoyy71GeKNoMewKKtoqWSYvjSb9iQMeQIIsS9u0CZA6zPQxjkE9J
6uynw8usu+HcHAKQ852r4hmrmgJZiuZalgGPVAfJbODgp89rJzhbVElwiwW77xHh
EYUET22yB00viK1zkIzIMtcPsNhcyTfxTYZIxI0DP/bqFBkpwAmTP3axwqmaoLAN
iHtInc+D4ajOTGWwzimQShQBvjPjeFvtf/ErqaS7Nj1IqGCXLu9nadt14Gf0OcgS
MTceMpLpPZVGGHJedlWtXrLFICSGFTNWsZQ7GcDQEPX8axNligCDtIZZ/PHhi5j9
Z52tD04emkuxI+mma80ejcnkKh4oI88dG9/GL4C8Qhua7MCJHzyASY7kmHZ7hqJB
Og8AYIXJVFHXw0wsgiIdhDQ3A4bwX31T8xc5lhpvSaZUJgnbih4QdA6oohHr4vRf
Z+Sx0DgksuscH8LIVgcJwcCLukX2Wb5IRvBiR3cSclQM20WEL66zt8ce15nMVECa
zO++FfcZ
=jSjm
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `1a5436e3da8426ac64d2e23be49dd4c3bbb869d0d35ecba05ef4c37aaa88e2a9  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401081a5436e3da8426ac64d2e23be49dd4c3bbb869d0d35ecba05ef4c37aaa88e2a9f01095fa19222b62741aedb159b0b961444b08fff01026953118e3a54afda3c9f530929fafb608f104602249ddf008f9ae20e6e08023250083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff0107253b5201d0c0190c5b873b718e617f808f104602257aff008401ed73015d02ade0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff01093424d4936e1aba210ff24738692779c08f020ac7e902439ee17b24f858ba8aa457232e1eef8fa736df9ec0b075a3848b5865908f104602257aff008e245fe8ba33098a40083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010c225143a4b6cd2f86f8f9b054d9fdae608f104602257aef008741c53665b9f26710083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

glozow commented:
> The third commit doesn't compile

Should be fixed now 🤦 thanks @MarcoFalke! Fixed the constructors a bit as well to take out the unnecessary `TxValidationState` in the success case and Assume not valid in the failure case.

MarcoFalke commented:
Changes since previous review:

* No longer pass redundant default-constructed state when valid
* Remove uint8_t from `enum class`
* Add `Assume(!state.IsValid());`
* Make it compile


ACK 53e716ea119658c28935fee24eb50090907c500e 💿

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 53e716ea119658c28935fee24eb50090907c500e 💿
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUh02Qv/VKTbpbQkoToYPmzKcSX5aJ+JAaQThI2/aGnpdgFnl/GdIVDyMcWHyygP
iduWhmF6BeEerKuctDAbfLB+EjeQdIr0xMa6lg5nkSsZTi0W+wn+5okOnngsjAzI
JazlKJJGQj/WjUJof04cTtIfcT9ecLS9yN8TJ/W8EWnm+RpSCOMBgUvRfBgdCBXX
GKs2xoxN6BT1ML5+y7qQQB2VyZE/DAEykbR6NhgaAIkrA5UZM6tP4t5/YTNtjoZq
hBgKkesPqym/WjB/6LTEwannfiW/ACvpM3Czg5pPS1svMyMxT4LQh4rMDiF4U21x
apPhPIMwvGce1pdX6UwMuDhRdA0VE6y23rxjSjhZaA0ZEBOVEW2sYBCa9bVcROrY
60WzJ2j6dYY2xIHJsYo5vG1LxjSrz42My50LLIKTBsihTg1LNst52KXZfFjozng6
2gewYrzP0lJe/fyUf9BxNK1w9+mbd5kCJDR1VHxsu9Yc5B+4gVKmva4uFodLd+Yo
Hyw2/p/q
=G4nn
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `5da03708a5e913024d3ccd093bac9c7a16f54f296f91a24b4c60d85316f124f4  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401085da03708a5e913024d3ccd093bac9c7a16f54f296f91a24b4c60d85316f124f4f010dbf4e15db0297abe2a684975de5b53bb08fff0103c8356ef40dfbff63cf3e580d783345808f0206b93b3d3d98c555731e6b0f49120e1fad86ef43f689b91f47662d7b082b3ade308f1046023c583f008e624396f57e356b40083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010b288e3c7814103fb47988aea3ec2891108f020ab86ad5622ca023a96192d5172d8353bf74e7a3da7c023972bf1cc76eeb9aa1208f1046023b7b2f0089e3aa532c8ca59d10083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010c7037a66c44947013768d62a496b4d9f08f020a2bc3a33ffc1ca271ee1829e534c9b997bfabeb62e1d8d99e40e9e188c52cbbb08f1046023c583f008b992adbbe2925ab50083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010e57588942800f2d3137b761fc592e9ba08f120f6a0a76abf1b18c99ca2d115f473bb28af4a931f921c241abc939064d07c3bf008f1046023c583f008ef52417ebb44d66f0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

MarcoFalke commented on src/validation.h:
```diff
@@ -181,12 +182,46 @@ void UnlinkPrunedFiles(const std::set<int>& setFilesToPrune);
+        VALID, //!> Fully validated, valid.
+        INVALID, //!> Invalid.
+    };
+    ResultType m_result_type;
```
nit: Would be nice if this was const, so that compilation fails instead of getting a valgrind error on runtime if this is unset.

MarcoFalke commented on src/validation.h:
```diff
@@ -181,12 +182,46 @@ void UnlinkPrunedFiles(const std::set<int>& setFilesToPrune);
+        INVALID, //!> Invalid.
+    };
+    ResultType m_result_type;
+    TxValidationState m_state;
```
same?

MarcoFalke commented on src/validation.h:
```diff
@@ -181,12 +182,46 @@ void UnlinkPrunedFiles(const std::set<int>& setFilesToPrune);
+
+    /** Constructor for success case */
+    explicit MempoolAcceptResult(std::list<CTransactionRef>&& replaced_txns, CAmount fees)
+        : m_result_type(ResultType::VALID), m_state(TxValidationState{}),
```
Haven't tried, but I think this can be written shorter

```suggestion
        : m_result_type(ResultType::VALID), m_state{},
```

MarcoFalke commented on src/validation.cpp:
```diff
@@ -556,12 +556,12 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
     std::vector<COutPoint>& coins_to_uncache = args.m_coins_to_uncache;
 
     // Alias what we need out of ws
+    TxValidationState &state = ws.m_state;
```
nit in 174cb5330af4b09f3a66974d3bae783ea43b190e: When touching the code it would be good it the correct style was used right away, instead of adding another commit afterward to fix it up once more. (This increases the `git blame` and `git log -S` depth)

glozow replied:
Mm, definitely - I will keep this in mind!

MarcoFalke commented on src/validation.cpp:
```diff
@@ -702,7 +699,7 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
 
     // nModifiedFees includes any fee deltas from PrioritiseTransaction
-    nModifiedFees = nFees;
+    nModifiedFees = ws.m_fee_out;
```
nit 174cb5330af4b09f3a66974d3bae783ea43b190e: When touching a member variable in all places where it is used it could make sense to use the right name right away instead of using the wrong name first in all sites and then adding a commit to rename it. (Increases git blame depth)

maflcko reviewed (commented):
Left some non-critical style nits

ariard reviewed (commented):
Code Review ACK 53e716e, I did tweak a bit the touched paths to see if we had good test coverage. Didn't find holes.

Nit: if you have to re-touch the code, you forget this one https://github.com/bitcoin/bitcoin/pull/21062#discussion_r569329362

jnewbery commented:
Code review ACK 53e716ea119658c28935fee24eb50090907c500e

glozow commented:
Opened followup #21146 to address style comments from @MarcoFalke and @ariard (oopsie for accidentally pushing here, please ignore that).
